File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function(heights) {
    // integer heights []
    // inputs: heights[] heights[i] is how tall it is , width is always 1
    // outputs: return largest rectangle ( area of rectangle )
    let maxArea = 0
    let stack = []
    heights.push(0)
    for (let i = 0; i < heights.length; i++) {
        let heightStart = i
        while (stack.length > 0 && stack[stack.length - 1][1] > heights[i]) {
            const [pos, height] = stack.pop()
            maxArea = Math.max(maxArea, height * (i - pos))
            heightStart = pos
        }
        stack.push([heightStart, heights[i]])

    }

    return maxArea 
};
// [[1,1] , [2,2]]

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # inputs: integers heights[]
        # return area of largest rectangle

        stack = [[0, heights[0]]]
        max_area = 0

        for idx, height in enumerate(heights):
            if height >= stack[-1][1]:
                stack.append([idx, height])
            else:
                while stack and height < stack[-1][1]:
                    i, h = stack.pop()
                    area = (idx - i) * h 
                    max_area = max(max_area, area)
                stack.append([i, height])


        for i, h in stack:
            width = len(heights) - i
            area = width * h
            max_area = max(max_area, area)

        return max_area

